src/rulesets/Base/array.jl

```diff
@@ -1,3 +1,40 @@
+#####
+##### constructors
+#####
+
+ChainRules.@non_differentiable (::Type{T} where {T<:Array})(::UndefInitializer, args...)
+
+function rrule(::Type{T}, x::AbstractArray) where {T<:Array}
+    project_x = ProjectTo(x)
+    Array_pullback(ȳ) = (NoTangent(), project_x(ȳ))
+    return T(x), Array_pullback
+end
+
+#####
+##### `vect`
+#####
+
+@non_differentiable Base.vect()
+
+# Don't worry about projection here. The data passes straight through, so if a cotangent has
+# the wrong type for some reason, it must be the fault of another rule somewhere.
+function rrule(::typeof(Base.vect), X...)
+    function vect_pullback(ȳ)
+        X̄ = ntuple(n -> ȳ[n], length(X))
```

Comment:
Move `N = length(X)` outside to avoid closing over X?

Reply:
Good idea.